### PR TITLE
[Bugfix] Fix gemma3 with transformers backend

### PR DESCRIPTION
**Pull request by zucchini-nlp**
Gemma3 had issues with transformers backend and was explicitly left out from supported models. There are two reason for that:

1. The model's embedding layer is a custom embedding module which applies scaling. vLLM replaces input embedding layer with internal class which doesn't handle scaling, so we were seeing garbage generation with the model
2. The model requires extra kwargs in the input, apart from text ids and images. To support bidirectional mask we need `token_type_ids`. I decided to exploit an existing `SupportsMultiModalWithRawInput` for that but keep input merging on vLLM side. If we were to delete merging and let transformers handle it, we will need a bigger PR. So lmk if the current appoach works or if you want me to delete input merging code

cc @hmellor @Isotr0py

### CODE REVIEW

**Review by Isotr0py — APPROVED** [review]

File: vllm/model_executor/models/transformers.py [hunk 1]
```diff
@@ -299,7 +299,7 @@ def _apply_hf_processor_text_mm(
         prompt_ids, = processed_data.pop("input_ids").tolist()
         mm_token_type_ids = processed_data.pop(
             "mm_token_type_ids"
-        ) if "mm_token_type_ids" in processed_data else processed_data.pop(
+        ) if "mm_token_type_ids" in processed_data else processed_data.get(
             "token_type_ids")  # for gemma3 only
```
[the remark this answers is not in the document]
Comment (DarkLight1337) [reply vllm/model_executor/models/transformers.py]: Looks like this comment is valid, based on the test failure

[the remark this answers is not in the document]
Comment (zucchini-nlp) [reply vllm/model_executor/models/transformers.py]: fixed, we need `token_type_ids` for gemma3 so we can't pop it here. Instead popping it from kwargs right where it failed

**Review by hmellor — APPROVED** [review]
LGTM let's wait until after https://github.com/vllm-project/vllm/pull/23749 to merge

**Comment by DarkLight1337** [comment]
There is also an issue related to `torch.compile` that is failing the CI in this PR

**Comment by zucchini-nlp** [comment]
Ah didn't check the tests assuming it was unrelated fails. Will check out
Update: hmm interesting that it doesn't fail for me locally

**Comment by hmellor** [comment]
Failure is relevant

**Comment by zucchini-nlp** [comment]
Yeah, and it seems to be fixed in 4.56.0 in transformers side. Let's wait for vLLM to dump the latest release then, otherwise I don't see way to fix it unless we skip compilation in gemma3 test case

**Comment by hmellor** [comment]
Sounds good, let's put this on hold until then.

File: vllm/model_executor/models/transformers.py [hunk 2]
```diff
@@ -482,6 +482,8 @@ def __init__(self, *, vllm_config: VllmConfig, prefix: str = ""):
         self.tensor_parallel()
 
         # Input embeddings
+        self.embed_scale = getattr(self.model.get_input_embeddings(),
+                                   "embed_scale", 1.0)
```
Comment (hmellor) [inline vllm/model_executor/models/transformers.py]: Where is this used? Can it go inside the `if` block below?

File: vllm/model_executor/models/transformers.py [hunk 3]
```diff
@@ -640,6 +642,8 @@ def create_attention_instances(
                 num_kv_heads=num_kv_heads,
                 cache_config=self.cache_config,
                 quant_config=self.quant_config,
+                logits_soft_cap=getattr(self.text_config,
+                                        "attn_logit_softcapping", None),
```
Comment (hmellor) [inline vllm/model_executor/models/transformers.py]: Could this have other names? If yes, you can use the `getattr_iter` util.

And can we put the `getattr` outside the loop? We should only need to get it once.

File: vllm/model_executor/models/transformers.py [hunk 4]
```diff
@@ -780,6 +786,7 @@ def _can_concat(x: list[torch.Tensor]):
     },
     enable_if=can_enable_torch_compile)
 class TransformersForMultimodalLM(TransformersForCausalLM, SupportsMultiModal):
+    supports_multimodal_raw_input_only = True
```
Comment (hmellor) [inline vllm/model_executor/models/transformers.py:852]: Is it safe to add this?

File: vllm/model_executor/models/transformers.py [hunk 5]
```diff
@@ -817,8 +824,10 @@ def forward(
         inputs_embeds: Optional[torch.Tensor] = None,
         **kwargs: object,
     ) -> Union[torch.Tensor, IntermediateTensors]:
+        kwargs = kwargs.get("token_type_ids") or {}
```
Comment (hmellor) [inline vllm/model_executor/models/transformers.py]: Maybe a comment to say that this is for Gemma 3 MM

File: vllm/model_executor/models/transformers.py (hunk 2 above)
Comment (zucchini-nlp) [reply vllm/model_executor/models/transformers.py]: It is used for gemma3 from LLMs and many enc-dec old models in transformers. I feel like the rebase deleted where it was used, it should be in `get_inputs_embeddings` as `inputs_embeds *= scale`

The if block is executed always right or is it only for specific cases when the weights are not tied?

File: vllm/model_executor/models/transformers.py (hunk 3 above)
Comment (zucchini-nlp) [reply vllm/model_executor/models/transformers.py]: i don't think so, haven't seen other names

File: vllm/model_executor/models/transformers.py (hunk 4 above)
Comment (zucchini-nlp) [reply vllm/model_executor/models/transformers.py:852]: yes, the only thing that the flag does in vLLM currently is to pass over all `mm_kwargs` as is to the forward. And then we forward we filter out what we need , whcih is the change from comment below

File: vllm/model_executor/models/transformers.py (hunk 2 above)
Comment (hmellor) [reply vllm/model_executor/models/transformers.py]: That if block is only executed on the first pipeline rank. So it should be safe to put it in the if block as `get_input_embeddings` should only be called on the first rank too?

File: vllm/model_executor/models/transformers.py [hunk 6]
```diff
@@ -824,7 +827,9 @@ def forward(
         inputs_embeds: Optional[torch.Tensor] = None,
         **kwargs: object,
     ) -> Union[torch.Tensor, IntermediateTensors]:
-        kwargs = kwargs.get("token_type_ids") or {}
+        # Gemma3 and PG needs `token_type_ids` to work correctly
```
Comment (hmellor) [inline vllm/model_executor/models/transformers.py]: PG -> PaliGemma?

Comment (zucchini-nlp) [reply vllm/model_executor/models/transformers.py]: yes

**Review by hmellor — APPROVED** [review]
LGTM! Thanks for fixing this model!

File: vllm/model_executor/models/transformers_moe.py [hunk 7]
```diff
@@ -263,6 +264,8 @@ class TransformersMoEForCausalLM(TransformersMoEBase, TransformersForCausalLM):
         "inputs_embeds": 0,
     },
     enable_if=can_enable_torch_compile)
-class TransformersMoEForMultimodalLM(TransformersMoEForCausalLM,
+class TransformersMoEForMultimodalLM(TransformersMoEBase,
                                      TransformersForMultimodalLM):
-    pass
+
+    def get_input_embeddings(self, *args, **kwargs) -> torch.Tensor:
```
Comment (DarkLight1337) [inline vllm/model_executor/models/transformers_moe.py]: `get_input_embeddings = SupportsMultiModal.get_input_embeddings` also works

Comment (hmellor) [reply vllm/model_executor/models/transformers_moe.py]: Thanks, this is a nicer way to do it!

**Comment by DarkLight1337** [comment]
@ywang96 since Transformers implements attention mask correctly, WDYT of removing the native vLLM implementation of gemma3?
